File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/interfaces/financial/PaymentAttributes.java
package com.emerchantpay.gateway.genesisandroid.api.interfaces.financial;

import com.emerchantpay.gateway.genesisandroid.api.models.Currency;
import com.emerchantpay.gateway.genesisandroid.api.util.RequestBuilder;

import java.math.BigDecimal;

public interface PaymentAttributes {

    RequestBuilder requestBuilder = new RequestBuilder("");

    // Payment Params
    PaymentAttributes setAmount(BigDecimal amount);

    BigDecimal getAmount();

    PaymentAttributes setCurrency(String currency);

    String getCurrency();

    default RequestBuilder buildPaymentParams() {

        BigDecimal convertedAmount = null;

        if (getAmount() != null && getCurrency() != null) {

            Currency curr = new Currency();

            curr.setAmountToExponent(getAmount(), getCurrency());
            convertedAmount = curr.getAmount();
        }

        requestBuilder.addElement("amount", convertedAmount)
                .addElement("currency", getCurrency());

        return requestBuilder;
    }
}

File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/interfaces/financial/PaymentAttributes.kt
package com.emerchantpay.gateway.genesisandroid.api.interfaces.financial

import com.emerchantpay.gateway.genesisandroid.api.models.Currency
import com.emerchantpay.gateway.genesisandroid.api.util.RequestBuilder

import java.math.BigDecimal

interface PaymentAttributes {

    val amount: BigDecimal?
        get() = null

    val currency: String?
        get() = ""

    fun buildPaymentParams(): RequestBuilder {

        var convertedAmount: BigDecimal? = null

        when {
            amount != null && currency != null -> {

                val curr = Currency()

                curr.setAmountToExponent(amount!!, currency!!)
                convertedAmount = curr.amount
            }
        }

        requestBuilder.addElement("amount", convertedAmount!!)
                .addElement("currency", currency!!)

        return requestBuilder
    }

    companion object {

        val requestBuilder = RequestBuilder("")
    }
}
